### PR TITLE
AWS Organizations - organizations__assume_role initial upload

### PR DESCRIPTION
**Pull request by WebbinRoot**
Added a module for AWS Organizations that will automatically check member accounts for roles to see if you can assume them. This represents lateral movement. As you will see, I was hoping to maybe create a key-set for each success if possible but wasn't sure how to program that into the current framework. At the moment the module just returns a list of roles you can assume. You can also supply a list of account numbers in a file or inline, and same with role names depending on what you want to do.

### CODE REVIEW

File: pacu/modules/organizations__assume_role/main.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
+#!/usr/bin/env python3
```
Comment (jdearmas) [inline pacu/modules/organizations__assume_role/main.py:1]: Please add a single comment at the top of the file explaining (at a high-level) what the code in the file does.

File: pacu/modules/organizations__assume_role/main.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    args = parser.parse_args(args)
+    print = pacu_main.print
+    fetch_data = pacu_main.fetch_data
+    ######
+
+    accounts = []
+    roles = ["OrganizationAccountAccessRole"]
+    
+    if args.accounts and args.accounts_file:
+        print("You must choose either accounts or the accounts-file argument or neither, you cannot choose both")
```
Comment (jdearmas) [inline pacu/modules/organizations__assume_role/main.py:102]: In order to enforce this, implement a [mutually exclusive group](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_mutually_exclusive_group) in argparse.

File: pacu/modules/organizations__assume_role/main.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    if args.roles and args.roles_file:
```
Comment (jdearmas) [inline pacu/modules/organizations__assume_role/main.py:103]: In order to enforce this, implement a [mutually exclusive group](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_mutually_exclusive_group) in argparse.

File: pacu/modules/organizations__assume_role/main.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        print("You must choose either roles or roles-file or neither, you cannot choose both")
+
+    # If user did not specify an account to try brute forcing, then try pre-requisite
+    if not args.accounts and not args.accounts_file:
+        if fetch_data(['Organizations'], 'organizations__enum', '--accounts') is False:
+            return
```
Comment (jdearmas) [inline pacu/modules/organizations__assume_role/main.py:109]: Add a print statement before the return keyword that says "Prerequisite module was not run. Exiting...".

File: pacu/modules/organizations__assume_role/main.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    # a module summary string.
+    data = roles_list
+    return data
+
+def summary(data, pacu_main):
+    if len(data) == 0:
+        return "No roles were found that could be assumed"
+    else:
+        summary = 'The following roles can be assumed as shown below:'
+        for vuln in data:
```
Comment (jdearmas) [inline pacu/modules/organizations__assume_role/main.py:148]: Change the vuln variable to 'arn'.

**Comment by jdearmas** [comment]
What do you think about changing the name of the module to be more general? Such as 'iam__brute_roles'? Since you can use this module to brute force any account, not just organizations enumerated accounts?

**Review by jdearmas — CHANGES_REQUESTED** [review]

**Comment by WebbinRoot** [comment]
I wouldn't be opposed to doing 'iam__brute_roles', the only tricky thing is pacu already has a kind of iam brute force module in the unauthenticated modules for iam__enum_roles. Originally I wanted this module to be pretty small and it would just call the iam brute force module in the background but that didn't seem like it could be done natively. So I brought this up March 31 in the Rhino Security Discord but I felt like that one should maybe be a bit more generalized, or maybe it could be this one. The last I heard:

"After some discussion, there's is a plan to add a flag to the iam__enum_roles module in order to skip the role policy update and go straight to brute forcing a list of assumed roles. Does this align with your question?"

There is just another module that brute forces roles in other AWS accounts so making that distinction somehow.  So I was making this module to be more organization-specific. I am flexible in whatever naming convention, just let me know what you want to do.

File: pacu/modules/organizations__assume_role/main.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+module_info = {
+    # Name of the module (should be the same as the filename).
+    'name': 'organizations__assume_role',
+
+    # Name and any other notes about the author.
+    'author': 'Scott (@WebbinRoot:/in/webbinroot/)',
+
+    # Category of the module. Make sure the name matches an existing category.
+    'category': 'LATERAL_MOVE',
```
Comment (jdearmas) [inline pacu/modules/organizations__assume_role/main.py:19]: The category needs to be per-defined or a new one must be defined.

File: pacu/modules/organizations__assume_role/main.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        #    )
+        #else:
+        #    pass
+
+    except ClientError as error:
+        code = error.response['Error']['Code']
+        if code == 'AccessDenied':
+            print("FAILURE: "+role_arn)
+        else:
+            print('(DEFAULT FAILURE: {}'.format(code))
```
Comment (jdearmas) [inline pacu/modules/organizations__assume_role/main.py:85]: There is an extra parenthesis in the print statement: '(DEFAULT ...'

File: pacu/modules/organizations__assume_role/main.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+    global roles_list
+    sts = boto3.client('sts')
+    try:
+        resp = sts.assume_role(
+            RoleArn=role_arn,
+            RoleSessionName=''.join(random.choice(string.ascii_letters) for _ in range(20)),
+        )
+        roles_list.append(role_arn)
+        print("SUCCESS: " + role_arn)
+
```
[the remark this answers is not in the document]
Comment (DaveYesland) [reply pacu/modules/organizations__assume_role/main.py:64]: I am not sure of a way to store creds like that either AFIK there is no built-in credential store. Maybe there should be..

But yeah if you think that will be too verbose leave as is.

File: pacu/modules/organizations__assume_role/main.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
+#!/usr/bin/env python3
+import argparse
+from botocore.exceptions import ClientError
+import boto3, random, string
+
+
+
+# When writing a module, feel free to remove any comments, placeholders, or
```
[the remark this answers is not in the document]
Comment (WebbinRoot) [reply pacu/modules/organizations__assume_role/main.py:8]: Fixed

File: pacu/modules/organizations__assume_role/main.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+    sts = boto3.client('sts')
+    try:
+        resp = sts.assume_role(
+            RoleArn=role_arn,
+            RoleSessionName=''.join(random.choice(string.ascii_letters) for _ in range(20)),
+        )
+        roles_list.append(role_arn)
+        print("SUCCESS: " + role_arn)
+        
+        # See if I can create a key set automatically in background
```
[the remark this answers is not in the document]
Comment (WebbinRoot) [reply pacu/modules/organizations__assume_role/main.py:65]: Fixed

File: pacu/modules/organizations__assume_role/main.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+    'arguments_to_autocomplete': ['accounts','accounts-file','roles','roles-file'],
+}
+
+parser = argparse.ArgumentParser(add_help=False, description=module_info['description'])
+
+parser.add_argument('--accounts', required=False, nargs='+', default=None,  help='Pass in a list of accounts to try assuming, --accounts 1 2 3')
+parser.add_argument('--accounts-file', required=False, nargs=1, default=None, help="Pass in a filename containing a list of accounts to try assuming")
+parser.add_argument('--roles', required = False, nargs = '+', default=None,  help='Pass in a list of roles to try assuming. If non are specified uses default OrganizationAccountAccessRole, --roles a b c')
+parser.add_argument('--roles-file', required=False, nargs=1, default=None, help='Pass in a filename containing a list of roles to try assuming')
+#parser.add_argument('--create-sessions', required=False, action="store_true", help='Will automatically create a session for every successful AssumeRole operation. Can check all of them after with swap_keys command.')
```
[the remark this answers is not in the document]
Comment (WebbinRoot) [reply pacu/modules/organizations__assume_role/main.py:50]: Fixed

File: pacu/modules/organizations__assume_role/main.py (hunk 5 above)
Comment (WebbinRoot) [reply pacu/modules/organizations__assume_role/main.py:148]: Fixed

File: pacu/modules/organizations__assume_role/main.py (hunk 4 above)
Comment (WebbinRoot) [reply pacu/modules/organizations__assume_role/main.py:109]: Fixed

File: pacu/modules/organizations__assume_role/main.py (hunk 2 above)
Comment (WebbinRoot) [reply pacu/modules/organizations__assume_role/main.py]: Fixed

File: pacu/modules/organizations__assume_role/main.py (hunk 3 above)
Comment (WebbinRoot) [reply pacu/modules/organizations__assume_role/main.py]: Fixed

File: pacu/modules/organizations__assume_role/main.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (WebbinRoot) [reply pacu/modules/organizations__assume_role/main.py]: Fixed

**Comment by WebbinRoot** [comment]
@jdearmas Made changes in latest commit. In terms of module name and long-term plans with assume-role brute forcing just let me know what decision is, flexible with whatever.

**Comment by DaveYesland** [comment]
@WebbinRoot looks like the STS client still needs to be fixed to use Pacu session creds.

File: pacu/modules/organizations__assume_role/main.py (hunk 7 above)
Comment (WebbinRoot) [reply pacu/modules/organizations__assume_role/main.py]: Fixed

**Comment by WebbinRoot** [comment]
@DaveYesland Ah I think I see, is this the use of the get_boto3_client method in Pacu? Changed it to that in the latest commits. I can change this in organizations__enum as well in another pull now that I see that.

**Review by DaveYesland — CHANGES_REQUESTED** [review]

**Review by DaveYesland — APPROVED** [review]
Thanks for the contribution, nice work on this one!

**Review by jdearmas — APPROVED** [review]

**Pull request merged**